Implement base structure for repository file artifacts

vaibhavdaren opened the pull request:
In this Pull request, I have added support for the Repository-File artifact type for the Almighty Install Plugin.
The classes add support for the file type. 
* [X] implement the feature
* [X] Initial Review   

For the next PR: https://github.com/teemtee/tmt/issues/3995#issuecomment-3349399405

happz commented:
@vaibhavdaren would it make sense to rebase this PR on top of the koji/brew one? https://github.com/teemtee/tmt/pull/4011 That one seems to be adding the basic structure, and koji/brew "artifacts", I would expect the repository file to build upon that, adding new artifact type, sharing the directories with koji/brew providers, and so on.

vaibhavdaren commented:
> @vaibhavdaren would it make sense to rebase this PR on top of the koji/brew one? #4011 That one seems to be adding the basic structure, and koji/brew "artifacts", I would expect the repository file to build upon that, adding new artifact type, sharing the directories with koji/brew providers, and so on.

I have rebased on top of his branch. Changed functions to match signatures. Let me review the orginal CR and put my thoughts on that as well.

vaibhavdaren commented:
From review meeting @happz @psss
Scope of this PR : Add a new artifact type `Repository-File` for almighty install plugin with base helpers.

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+            guest.execute(
+                ShellScript(
+                    f"curl -L --fail -o {quote(str(destination))} {quote(url)}"
+                ).to_shell_command(),
```
Can safely remove `to_shell_command()`

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+            # Ensure temporary download directory exists on guest
+            guest.execute(
+                ShellScript(
+                    f"[ -d {quote(str(destination.parent))} ] || "
```
This directory creation logic is redundant since the base class `download_artifacts()` method already ensures the destination directory exists before calling `_download_artifact()`. This entire block can be removed.

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+        :yields: A RepositoryArtifactInfo instance for the .repo file
+        """
+        self.logger.debug(f"Listing repository artifact for URL: {self.artifact_id}")
+        yield RepositoryArtifactInfo(url=self.artifact_id)
```
This can be cached as this will jsut re-validate the same url each time.
I just refactored my [PR](https://github.com/teemtee/tmt/pull/4011/commits/f49b686de951f4c53cecaeb0a1c175607c236925) too :)

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+        """
+        parsed_url = urlparse(self.url)
+        filename = Path(parsed_url.path).name
+        if not filename.endswith('.repo'):
```
Swap the order of these two checks for clearer error messages.
For example, a URL like `https://example.com/` would show "URL must point to a .repo file, got: " instead of "Invalid filename extracted from URL".

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+
+        parsed_url = urlparse(artifact_id)
+        filename = Path(parsed_url.path).name
+        if not filename.endswith('.repo'):
```
Again the order to be swapped.

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+        :returns: The validated URL
+        :raises ValueError: If the URL is invalid or does not point to a .repo file
+        """
+        if not artifact_id.startswith(('http://', 'https://')):
```
A question: can someone point to a local repo file?

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+            self.logger.debug(f"Downloading {filename} to temporary {destination}")
+            guest.execute(
+                ShellScript(
+                    f"curl -L --fail -o {quote(str(destination))} {quote(url)}"
```
Maybe `--insecure` instead?

AthreyVinay commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+        :param url: URL to the .repo file
+        :param kwargs: Additional arguments for base class
+        """
+        super().__init__(id=0, _raw_artifact={}, **kwargs)  # Dummy id and empty _raw_artifact
```
Assune this will need to have a real / unique `id` and `_raw_artifact`

happz commented:
Bumped from 1.57, will be reviewed on the planning meeting.

vaibhavdaren commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,164 @@
+            guest.execute(
+                ShellScript(
+                    f'{"sudo " if not guest.facts.is_superuser else ""}'
+                    f"mv {quote(str(destination))} {quote(str(repo_destination))} && "
```
Added in the next commit.

LecrisUT reviewed (commented):
Just 2 remaining comments, and the logging issue. I am fine with deferring the logging issue to later when we can test it in context. Other than that lgtm.

Maybe one improvement would be to download directly in `/etc/yum.repos.d` but that needs some interface refactoring at the top level, so that can be deferred as well.

vaibhavdaren commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,170 @@
+                guest.execute(ShellScript(f"test -f {quote(str(repo_destination))}"), silent=True)
+                self.logger.warning(f"Overwriting existing repository file: {repo_destination}")
+            except GeneralError:
+                pass  # File does not exist, no need to warn
```
I got a comment in PR regarding this : [here](https://github.com/teemtee/tmt/pull/4018#discussion_r2335029752)
An artifact is either correctly placed / downloaded or not. I think we dont need to bother in between states. 
But the only thing I can think of is  this can only happen in the case if two repositories have the same name which is unlikely or previous failed run on a host.

LecrisUT replied:
Yes, but if we/the user installs the specific repo, it is done so intentionally. Sure it could catch some unintentional overwrites for ssh provisions or such, but you would have various other problems with those scenarios then.

One thing that I would consider is that these warnings are not yet properly forwarded to testing-farm so they would still be likely ignored. But again another issue to be resolved more broadly

psss commented on tmt/steps/prepare/brand_new_allmighty_install/providers/__init__.py:
```diff
@@ -18,6 +18,7 @@
 class ArtifactType(enum.Enum):
     RPM = 'rpm'
     CONTAINER = 'container'
+    REPOSITORY_FILE = 'repository-file'
```
Is repository file a type of an artifact? I thought artifacts are the bits to be tested (e.g. package) not the channel through which we get them (repo, repo file...).

psss commented on tmt/steps/prepare/brand_new_allmighty_install/providers/repository_provider.py:
```diff
@@ -0,0 +1,141 @@
+
+
+@container
+class RepositoryArtifactInfo(ArtifactInfo):
```
I thought that `ArtifactInfo` is supposed to be used for information about the artifacts, e.g. `rpms`, not information about the delivery method, e.g. repo file. Let's clarify these on the review session?

LecrisUT commented on tmt/steps/prepare/artifact/providers/repository_provider.py:
```diff
@@ -0,0 +1,115 @@
+    _raw_artifact: str
+
+    def __post_init__(self) -> None:
+        """
+        Validate the repository file URL after initialization.
+
+        Raises:
+            ValueError: If the URL scheme is not HTTP/HTTPS or if the file
+                        does not have a ``.repo`` extension.
+        """
```
`__post_init__` is a non-exposed method, it does not really need a docstring. But it is also in the wrong format.

LecrisUT commented on tmt/steps/prepare/artifact/providers/repository_provider.py:
```diff
@@ -0,0 +1,115 @@
+"""
+This file provides classes for handling repository file artifacts in the tmt framework.
+It includes the `RepositoryFileInfo` class, which encapsulates metadata about a repository
+file (e.g., URL and filename), and the `RepositoryFileProvider` class, which manages the
+downloading of `.repo` files to a guest system.
+"""
```
If the docstring is short and concise I don't mind having them for module. But statements like "This file provides ..." put me off because it is a description that is self-evident. Instead something like
```python
"""
Artifact providers for Repository and Repository files.
"""
```

Yes, that statement is also self-evident and merely fluff, but the user does not have to read full paragraphs to find that out.

LecrisUT commented on tmt/steps/prepare/artifact/providers/repository_provider.py:
```diff
@@ -0,0 +1,115 @@
+        return self._raw_artifact
+
+
+class RepositoryFileProvider(ArtifactProvider[RepositoryFileInfo]):
```
Note that this should be converted to `ArtifactProvider[RpmArtifactInfo]` because we will be listing RPMs. It is unclear without the user part of it since the whole `download_artifacts` needs equivalent adjustments, but at least put a TODO note about that.

vaibhavdaren replied:
Added Todo in [commit](https://github.com/teemtee/tmt/pull/4018/commits/fd34ba18d1c80ac7be30d79d7a69f559ab9396f9).

LecrisUT reviewed (approved):
Fine with me in this form, mostly because all of the other changes that would make sense to make here require having the context of the `Prepare` plugin to actually make sense *how* we would need to adjust things.

tcornell-bus reviewed (approved):
LGTM, unit tests would be nice, but those can always be added later.

thrix commented on tmt/steps/prepare/artifact/providers/repository_provider.py:
```diff
@@ -0,0 +1,108 @@
+                f"Invalid repository URL: {self._raw_artifact}. Only HTTP(S) is supported."
+            )
+        path = tmt.utils.Path(parsed_url.path)
+        if path.suffix != '.repo':
```
I am not sure we need to be here so strict, what if user has a URL with query params ...

```
pathlib.Path("https://abc.com/what.repo?wtf").suffix == ".repo?wtf"
```

thrix commented on tmt/steps/prepare/artifact/providers/repository_provider.py:
```diff
@@ -0,0 +1,108 @@
+        try:
+            # Download the file to the temporary destination
+            guest.execute(
+                ShellScript(f"curl -L --fail -o {quote(str(destination))} {quote(url)}"),
```
We need to add retries here. Unfortunatley curl does not support good retrying for older versions, so some other way of retries might be needed. Or if we can know the guest is a more recent guest and fallback, dunno

vaibhavdaren commented:
Doc tests are failing here 
```
['Traceback (most recent call last):\n', '  File "/home/runner/work/tmt/tmt/.hatch_envs/lib/python3.13/site-packages/sphinx/ext/autodoc/importer.py", line 269, in import_object\n    module = import_module(modname, try_reload=True)\n', '  File "/home/runner/work/tmt/tmt/.hatch_envs/lib/python3.13/site-packages/sphinx/ext/autodoc/importer.py", line 175, in import_module\n    module = importlib.import_module(modname)\n', '  File "/opt/hostedtoolcache/Python/3.13.7/x64/lib/python3.13/importlib/__init__.py", line 88, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import\n', '  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load\n', '  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked\n', '  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked\n', '  File "<frozen importlib._bootstrap_external>", line 1026, in exec_module\n', '  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed\n', '  File "/home/runner/work/tmt/tmt/tmt/steps/prepare/artifact/providers/repository_provider.py", line 15, in <module>\n    from tmt.steps.prepare.artifact.providers.info import RpmArtifactInfo\n', "ModuleNotFoundError: No module named 'tmt.steps.prepare.artifact.providers.info'\n"] [autodoc.import_object]
```
Let me check

AthreyVinay commented:
> Doc tests are failing here
> 
> ```
> ['Traceback (most recent call last):\n', '  File "/home/runner/work/tmt/tmt/.hatch_envs/lib/python3.13/site-packages/sphinx/ext/autodoc/importer.py", line 269, in import_object\n    module = import_module(modname, try_reload=True)\n', '  File "/home/runner/work/tmt/tmt/.hatch_envs/lib/python3.13/site-packages/sphinx/ext/autodoc/importer.py", line 175, in import_module\n    module = importlib.import_module(modname)\n', '  File "/opt/hostedtoolcache/Python/3.13.7/x64/lib/python3.13/importlib/__init__.py", line 88, in import_module\n    return _bootstrap._gcd_import(name[level:], package, level)\n           ~~~~~~~~~~~~~~~~~~~~~~^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^\n', '  File "<frozen importlib._bootstrap>", line 1387, in _gcd_import\n', '  File "<frozen importlib._bootstrap>", line 1360, in _find_and_load\n', '  File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked\n', '  File "<frozen importlib._bootstrap>", line 935, in _load_unlocked\n', '  File "<frozen importlib._bootstrap_external>", line 1026, in exec_module\n', '  File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed\n', '  File "/home/runner/work/tmt/tmt/tmt/steps/prepare/artifact/providers/repository_provider.py", line 15, in <module>\n    from tmt.steps.prepare.artifact.providers.info import RpmArtifactInfo\n', "ModuleNotFoundError: No module named 'tmt.steps.prepare.artifact.providers.info'\n"] [autodoc.import_object]
> ```
> 
> Let me check

Module `info.py` has been removed :)

happz commented on tmt/steps/prepare/artifact/providers/repository.py:
```diff
@@ -0,0 +1,131 @@
+    @property
+    def filename(self) -> str:
+        """A suitable filename extracted from the URL path."""
+        path = urlparse(self.url).path
```
Since `urlparse()` is called in `__init__()`, maybe we could save its output, e.g. in `self._url_components`, and just use `self._url_components.path`.

vaibhavdaren replied:
Addressed in https://github.com/teemtee/tmt/pull/4018/commits/309d42757a5a75ebec1f8cdafcca7aca135b357c

happz commented on tmt/steps/prepare/artifact/providers/repository.py:
```diff
@@ -0,0 +1,131 @@
+        Validate the artifact identifier using the RepositoryFile class.
+        """
+        # The constructor of RepositoryFile handles URL validation
+        RepositoryFile(url=artifact_id)
```
The exactly same object has been already assigned to `self.repo_file`. Since that attribute is set, we are free to do nothing here.

vaibhavdaren commented on tmt/steps/prepare/artifact/providers/repository.py:
```diff
@@ -0,0 +1,131 @@
+    @property
+    def url(self) -> str:
+        """The URL of the .repo file."""
+        return self._url
```
Addressed in https://github.com/teemtee/tmt/pull/4018/commits/309d42757a5a75ebec1f8cdafcca7aca135b357c

vaibhavdaren commented on tmt/steps/prepare/artifact/providers/repository.py:
```diff
@@ -0,0 +1,131 @@
+    ) -> None:
+        """This provider only sets up the repo, it does not download RPMs."""
+
+    def download_artifacts(
```
I think each provider should have a common function `provide` `fetch` better named then `download_artifacts`. 
When the provider is Koji/Brew it really downloads the packages but when the provider is `repository` it just makes the repo file available. 

Lets discuss this in detail tomorrow. But for each of the providers `Provider.provide()`  or `Provider.fetch()`  should be common.

happz commented on tmt/steps/prepare/artifact/providers/repository.py:
```diff
@@ -0,0 +1,131 @@
+    def filename(self) -> str:
+        """A suitable filename extracted from the URL path."""
+        path = self._parsed_url.path
+        return unquote(Path(path).name)
```
How about `return unquote(Path(self._parsed_url.path).name)`? `path` variable seems very needless.
